motionplan/utils.go

```diff
@@ -98,3 +45,245 @@ func (r *resultPromise) result() ([]node, error) {
+	if err != nil {
+		return nil, err
+	}
+	if pivotFrame.Name() == referenceframe.World {
+		frames = uniqInPlaceSlice(append(solveFrameList, goalFrameList...))
+		moving, err = movingFS(solveFrameList)
+		if err != nil {
+			return nil, err
+		}
+		movingSubset2, err := movingFS(goalFrameList)
+		if err != nil {
+			return nil, err
+		}
+		if err = moving.MergeFrameSystem(movingSubset2, moving.World()); err != nil {
+			return nil, err
+		}
+	} else {
+		dof := 0
+		var solveMovingList []referenceframe.Frame
+		var goalMovingList []referenceframe.Frame
+
+		// Get minimal set of frames from solve frame to goal frame
+		for _, frame := range solveFrameList {
+			if frame == pivotFrame {
+				break
+			}
+			dof += len(frame.DoF())
+			frames = append(frames, frame)
+			solveMovingList = append(solveMovingList, frame)
+		}
+		for _, frame := range goalFrameList {
+			if frame == pivotFrame {
+				break
+			}
+			dof += len(frame.DoF())
+			frames = append(frames, frame)
+			goalMovingList = append(goalMovingList, frame)
+		}
+
+		// If shortest path has 0 dof (e.g. a camera attached to a gripper), translate goal to world frame
+		if dof == 0 {
+			worldRooted = true
+			frames = solveFrameList
+			moving, err = movingFS(solveFrameList)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			// Get all child nodes of pivot node
+			moving, err = movingFS(solveMovingList)
+			if err != nil {
+				return nil, err
+			}
+			movingSubset2, err := movingFS(goalMovingList)
+			if err != nil {
+				return nil, err
+			}
+			if err = moving.MergeFrameSystem(movingSubset2, moving.World()); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return &motionChain{
+		movingFS:       moving,
+		frames:         frames,
+		solveFrameName: solveFrame.Name(),
+		goalFrameName:  goalFrame.Name(),
+		worldRooted:    worldRooted,
+	}, nil
+}
+
+// findPivotFrame finds the first common frame in two ordered lists of frames.
+func findPivotFrame(frameList1, frameList2 []referenceframe.Frame) (referenceframe.Frame, error) {
+	// find shorter list
+	shortList := frameList1
+	longList := frameList2
+	if len(frameList1) > len(frameList2) {
+		shortList = frameList2
+		longList = frameList1
+	}
+
+	// cache names seen in shorter list
+	nameSet := make(map[string]struct{}, len(shortList))
+	for _, frame := range shortList {
+		nameSet[frame.Name()] = struct{}{}
+	}
+
+	// look for already seen names in longer list
+	for _, frame := range longList {
+		if _, ok := nameSet[frame.Name()]; ok {
+			return frame, nil
+		}
+	}
+	return nil, errors.New("no path from solve frame to goal frame")
+}
+
+// uniqInPlaceSlice will deduplicate the values in a slice using in-place replacement on the slice. This is faster than
+// a solution using append().
+// This function does not remove anything from the input slice, but it does rearrange the elements.
+func uniqInPlaceSlice(s []referenceframe.Frame) []referenceframe.Frame {
```

Comment:
can we rename this to uniqueFrameSlice or similar?

Reply:
This is the same name the function had in the solver frame file, just moved.

The reason for this name is to make it clear that the passed in array is getting mutated.